rust/lance-index/src/vector/hnsw/builder.rs

```diff
@@ -204,6 +206,46 @@ impl HNSW {
         Ok(hnsw)
     }
 
+    pub fn build_empty(distance_type: DistanceType, params: HnswBuildParams) -> Self {
+        let inner = HNSWBuilderInner::with_params(distance_type, params, 0);
+        Self {
+            inner: Arc::new(inner),
+        }
+    }
+
+    pub fn insert(
+        &self,
+        node: u32,
+        visited_generator: &mut VisitedGenerator,
+        storage: &impl VectorStore,
+    ) -> Result<()> {
+        if node as usize >= self.inner.nodes().read().unwrap().len() {
+            let mut write_nodes = self.inner.nodes.write().unwrap();
+            if node as usize >= write_nodes.len() {
+                let mut new_nodes: Vec<_> = write_nodes
+                    .iter()
+                    .map(|v| RwLock::new(v.read().unwrap().clone()))
```

Comment:
nit: could be
```
write_nodes.iter().map(|v| RwLock::new(v.read().unwrap().clone())).chain(
    write_nodes.len()..new_size.map(|i| ...)
)
```
this way the `new_nodes` vec will have to correct capacity in one allocation.